File: crates/guide-core/src/models/campaign.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::GameSystem;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Campaign {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub game_system: GameSystem,
    pub world_state: Option<WorldState>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// High-level persistent world metadata attached to a campaign.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorldState {
    pub current_location: Option<String>,
    pub current_date_in_world: Option<String>,
    pub active_quests: Vec<String>,
    pub completed_quests: Vec<String>,
    pub custom_notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateCampaignRequest {
    pub name: String,
    pub description: Option<String>,
    pub game_system: Option<GameSystem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateCampaignRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub world_state: Option<WorldState>,
}

File: src/guide/models/campaign.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from .shared import GameSystem


class WorldState(BaseModel):
    current_location: str | None = None
    current_date_in_world: str | None = None
    active_quests: list[str] = Field(default_factory=list)
    completed_quests: list[str] = Field(default_factory=list)
    custom_notes: str | None = None


class Campaign(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    game_system: GameSystem = GameSystem.dnd5e
    world_state: WorldState | None = None
    created_at: datetime
    updated_at: datetime


class CreateCampaignRequest(BaseModel):
    name: str
    description: str | None = None
    game_system: GameSystem | None = None


class UpdateCampaignRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    world_state: WorldState | None = None
